src/main/java/ee/bitweb/core/trace/thread/decorator/SecurityAwareMDCTaskDecorator.java

```diff
@@ -0,0 +1,35 @@
+package ee.bitweb.core.trace.thread.decorator;
+
+import ee.bitweb.core.trace.thread.ThreadTraceIdResolver;
+import lombok.RequiredArgsConstructor;
+import org.slf4j.MDC;
+import org.springframework.core.task.TaskDecorator;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+import java.util.Map;
+
+@RequiredArgsConstructor
+public class SecurityAwareMDCTaskDecorator implements TaskDecorator {
+
+    private final ThreadTraceIdResolver resolver;
+
+    @Override
+    public Runnable decorate(Runnable runnable) {
+        Map<String, String> contextMap = MDC.getCopyOfContextMap();
+        SecurityContext securityContext = SecurityContextHolder.getContext();
+
+        return () -> {
+            try {
+                MDC.setContextMap(contextMap);
+                SecurityContextHolder.setContext(securityContext);
```

Comment:
Even though SecurityContextHolder states that null is never given back, maybe null check should be added as well as setContext never expects null as value. This would serve as extra check.